app/src/main/java/ca/vijayan/flyweb/proxy/ProxyServer.java

```diff
@@ -266,15 +286,102 @@ void handleServiceConnect(ProxyConnection conn) {
+        final int BUFFER_SIZE = 1024;
+        ByteBuffer buf = ByteBuffer.allocate(BUFFER_SIZE);
+        int totalRead = 0;
+        for (;;) {
+            int nread = 0;
+            try {
+                nread = conn.getServiceSocket().read(buf);
+            } catch (IOException exc) {
+                Log.e("ProxyServer", "Failed to read from service connection.");
+                conn.markServiceSocketError();
+                break;
+            }
+            if (nread <= 0) {
+                break;
+            }
+            buf.flip();
+            conn.writeServiceRecvBuffer(buf);
+            totalRead += nread;
+            if (nread < BUFFER_SIZE) {
+                break;
+            }
+
+            buf.reset();
         }
-        conn.markServiceSocketConnected();
+
+        if (totalRead > 0) {
+            // Handle new service incoming data.
+            handleServiceIncomingData(conn, totalRead);
+        }
+    }
+
+    void handleServiceIncomingData(final ProxyConnection conn, final int nbytes) {
+        ByteBuffer buf = conn.getServiceRecvBuffer();
+        byte[] data = new byte[buf.remaining()];
+        buf.get(data);
+
+        ExecutorService executor = Executors.newFixedThreadPool(NUM_FIXED_THREAD_POOL);
+        final byte[] DATA = data;
+
+        Future<?> future = executor.submit(new Callable<Void>() {
+            public Void call() {
+                mProxyDataHandler.handleServiceDataReceived(conn, DATA);
+                return null;
+            }
+        });
+
+        getFromFuture(future, "Failed to handle service incoming data.");
     }
+
     void handleServiceWrite(ProxyConnection conn) {
+        int totalWritten = 0;
+        for (;;) {
+            if (!conn.getServiceSendBuffer().hasRemaining()) {
+                // Nothing to write.
+                return;
+            }
+
+            int nwritten;
+            try {
+                nwritten = conn.getServiceSocket().write(conn.getServiceSendBuffer());
+            } catch (IOException exc) {
+                Log.e("ProxyServer", "Failed to write to service connection.");
+                conn.markServiceSocketError();
+                break;
+            }
+
+            if (nwritten <= 0) {
+                break;
+            }
+            totalWritten += nwritten;
+        }
+
+        if (totalWritten > 0) {
+            handleServiceOutgoingData(conn, totalWritten);
+        }
+    }
+
+    void handleServiceOutgoingData(final ProxyConnection conn, final int nbytes) {
+        ExecutorService executor = Executors.newFixedThreadPool(NUM_FIXED_THREAD_POOL);
+        Future<?> future = executor.submit(new Callable<Void>() {
+            public Void call() {
+                mProxyDataHandler.handleServiceDataSent(conn, nbytes);
+                return null;
+            }
+        });
+
+        getFromFuture(future, "Failed to handle local outgoing data.");
+    }
+
+    void getFromFuture(Future future, String logMsg) {
```

Comment:
This method should no longer be necessary.  We fire and forget the callables to the executdor.